### DISCUSSION TITLE
Map all Dogecoin nodes!

### DISCUSSION THREAD

**Discussion by qlpqlp** [dogecoin/dogecoin#2636] [Help needed]
Hello,

Im trying to map all dogecoin nodes, and trying to not use the nmap script to parse each IP to find live nodes :)

I know that the command getpeerinfo lists the current connected nodes, but it dosent allow to renew the list or get all available nodes.

Bitcoin has a command getaddr but it sems is not included on 1.14.4, didnt check on 1.21 yet :)

Is there an easy way to get all IP's without the risk of being flaged as a DDoS ?

My hope is to mantain an active list on GIT of all Dogecoin nodes, and also to build a public DB to all.

Thanks in advance :)

**Comment by qlpqlp** [comment]
Hello again :)

So I managed to get around this "limitation"  of getting all nodes :)

Using the RCP's commands `getpeerinfo ` and `setban ` Im getting now all nodes.
I have coded an PHP script on my local dev enviorment to connect to all my Dogenodes using RCP and everyting is working great thanks to @patricklodder tips on RCP connection :)

The script collects all current peers connected on my Nodes using `getpeerinfo `, after that using JSON I collect all info and populate an MariaDB table,

Then I `setban ` all collected peers, and the Doge Node DNS 's get more peers to collect them :)

Hope to build contact DB and share to all on GIT the Code and DB wen it reachs the 1,864 Nodes that are displying now on https://blockchair.com/en/dogecoin/nodes

The DB will also include Country/City using GEO DB, but like me, I live in Portugal and have several servers on Hetzner located in Germany and OVH, located in France lol, also some nodes on my garage eheh :P

Also I have a script alredy coded that will try to find the IP info using Search Engines like Google and Social Media API's to parse info to try to collect contacts (email, phones numbers, names)

![FCOlzM_VIAMAMas](https://user-images.githubusercontent.com/5968076/138295823-af79e93d-d489-43b8-be1d-d9d373eefb62.jpg)

**Reply by qlpqlp** [reply]
Strange nodes showing in. Why no hard coded version or subver? forked? @patricklodder @rnicoll @michilumin 

Found almost 500 nodes until now, still working :)

![Screenshot_20211021-182019_Remote Desktop.jpg](https://user-images.githubusercontent.com/5968076/138327732-aec9b3cf-49de-4534-ae8e-d06899a0c2eb.jpg)

**Reply by rnicoll** [reply]
It's open source, so yes the version is compiled in but equally yes anyone can change it trivially.

That said, I think zero is a placeholder until version handshake has occurred. The nodes disconnect any node who advertises a version that's too low (check `src/version.h` for numbers).

**Reply by qlpqlp** [reply]
One more question to @patricklodder @rnicoll @michilumin ,

Does the getpeerinfo, on the Peer Index, "id" field, is this id unic? or is regenerating to anouther node if a previous node is offline?
Also see that is related to Peer version field somehow, but not sure :)

In outher words, is there any way to real count unic nodes? or is just a speculative guess like the crypto market? lol  :)

Beucause my script is geting some repeted Peer Index id's with diferent IP's (sure, can be the dinamic IP from the ISP), but just to make sure :)

**Reply by rnicoll** [reply]
I believe `id` is an index into an in-memory array, so definitely not unique. IP & port is a better identifier, but nodes intentionally don't provide a unique identifier (because it would impact anonymity)

**Reply by qlpqlp** [reply]
Many thanks @rnicoll , and sorry from some of the questions being to basic :)

I only started on blockchain in May 2021 lol because, finally I had some "free" time from one of my company brands that I sold in 2020 (thats why I detected all copycats brands and informed the Dogecoin Foundation in the past), and also my 2 kids that are now more "independent", so a lot of reading, a lot of youtube videos (including your streams eheh) and again a lot of reading to try to understand how everything works with blockchain and how can i help improve on my side :)

keep it going to al Dev Shibs! :) 
Everytime i wake up, now I have always something to do lol, help Dogecoin adoption and improvment!

**Comment by qlpqlp** [comment]
Here u go a preview of my mapping :) https://twitter.com/inevitable360/status/1451551241344102402

For this I used:

- Dogecoin Core Wallet with RCP enable to capture all Peers's (still in progress, almost 1000 nodes)
- PHP + MariaDB to interact with Dogecoin RCP
- Done a funny GIF Doge animation for "minning" movment
- Maxmind GeoIP Lite 2 to cross with Peers IP location
- OpenLayers for Mapping on the World Map from OpenStreetMap

To do:
- Identifiy by diferent colors the Doge nodes versions
- PING all IP's, from time to time to see if they are still alive, or if there are new ones.
- Reference contacts to each node
- Clean code and add to my GIT to all to use it

**Reply by qlpqlp** [reply]
Also will map all Dogecoin buisness's like the nodes , but will open anouther discussion :)

**Comment by PolinaShanti** [comment]
Hey there!
If you're looking for a free API key for Doge, please, take a look at https://nownodes.io/nodes/doge

NOWNodes is a blockchain-as-a-service solution that lets users get access to full Nodes and blockbook Explorers via API. Instant access to 56 blockchain networks, including BTC, ETH, and BSC, which has over 1000 coins and tokens.
